features/cellular/framework/common/CellularList.h

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (c) 2017, Arm Limited and affiliates.
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CELLULAR_LIST_H_
+#define CELLULAR_LIST_H_
+
+#include <stddef.h>
+
+namespace mbed {
+
+/** Class CellularList
+ *
+ *  Templated linked list class for common usage.
+ *
+ */
+template <class T> class CellularList
```

Comment:
no any std container instead of this new class?

Reply:
It seems that STL is not used anywhere in mbed-os... Do you think STL is safe to use?

Reply:
Would be interested in the answer to this - I was under the impression that even simple STL use tended to pull in far too much C++ library code on some toolchains - only seen some use in applications.

If you turn on FEATURE_COMMON_PAL, ns_list.h in there is solid and efficient - main advantage over STL is that no dynamic memory allocation is built-in, so no list operations can fail. Syntax a bit more clunky due to being C  - magic macros. But will definitely require less ROM due to type erasure rather than plain templates.

Either way, I don't think a change is worthwhile at this stage - this is basically an internal implementation detail. We need to concentrate on externally visible stuff to get this done.

Reply:
FWIW, I note that https://os.mbed.com/docs/v5.7/reference/guidelines.html does explicitly say "Avoid the STL due to unknown effect on system resources."